### PR TITLE
Finer-grained STOP effector parameters.

### PR DESCRIPTION
**Pull request by neykov**
Separate flags for process and machine stop, better control with regard to the machine state.

### CODE REVIEW

File: software/base/src/main/java/brooklyn/entity/basic/SoftwareProcess.java [hunk 1]
```diff
@@ -261,6 +261,20 @@ private ChildStartableMode(boolean isDisabled, boolean isBackground, boolean isL
...
+        public enum StopMode { ALWAYS, IF_NOT_STOPPED, NEVER };
+
+        @Beta /** @since 0.7.0 semantics of parameters to restart being explored */
+        public static final ConfigKey<StopMode> STOP_PROCESS_MODE = ConfigKeys.newConfigKey(StopMode.class, "stopProcessMode",
+                "When to stop the process with regard to the entity state", StopMode.IF_NOT_STOPPED);
+
+        @Beta /** @since 0.7.0 semantics of parameters to restart being explored */
+        public static final ConfigKey<StopMode> STOP_MACHINE_MODE = ConfigKeys.newConfigKey(StopMode.class, "stopMachineMode",
+                "When to stop the machine with regard to the entity state. " +
+                "ALWAYS will try to stop the machine even if the entity is already stopped," +
```
Comment (sjcorbett) [inline software/base/src/main/java/brooklyn/entity/basic/SoftwareProcess.java]: Add a space at the end of the line.

File: software/base/src/main/java/brooklyn/entity/basic/SoftwareProcess.java [hunk 2]
```diff
@@ -261,6 +261,20 @@ private ChildStartableMode(boolean isDisabled, boolean isBackground, boolean isL
...
+                "IF_NOT_STOPPED stops the machine only if entity not already stopped, " +
+                "NEVER doesn't stop the machine. ", StopMode.IF_NOT_STOPPED);
```
Comment (sjcorbett) [inline software/base/src/main/java/brooklyn/entity/basic/SoftwareProcess.java]: Trailing space.

File: software/base/src/main/java/brooklyn/entity/software/MachineLifecycleEffectorTasks.java [hunk 3]
```diff
@@ -76,6 +77,7 @@
 import brooklyn.util.text.Strings;
 import brooklyn.util.time.Duration;
 
+import com.google.api.client.repackaged.com.google.common.annotations.VisibleForTesting;
```
Comment (sjcorbett) [inline software/base/src/main/java/brooklyn/entity/software/MachineLifecycleEffectorTasks.java]: Use `com.google.common.annotations.VisibleForTesting` instead.

File: software/base/src/main/java/brooklyn/entity/software/MachineLifecycleEffectorTasks.java [hunk 4]
```diff
@@ -614,6 +631,23 @@ public void stop(ConfigBag parameters) {
         if (log.isDebugEnabled()) log.debug("Stopped software process entity "+entity());
     }
 
+    @VisibleForTesting
+    public static boolean canStop(StopMode stopMode, boolean isEntityStopped) {
```
Comment (sjcorbett) [inline software/base/src/main/java/brooklyn/entity/software/MachineLifecycleEffectorTasks.java]: May as well reduce the visibility of the method if it's only exposed for testing.

File: software/base/src/test/java/brooklyn/entity/basic/SoftwareProcessEntityTest.java [hunk 5]
```diff
@@ -267,6 +275,87 @@ public void testBasicSoftwareProcessStopsProcess() throws Exception {
...
+                StopMode.IF_NOT_STOPPED, false, true,
+                StopMode.NEVER, true, false,
+                StopMode.NEVER, false, false
+        };
+        for (int i = 0; i < matrix.length; i+=3) {
+            StopMode mode = (StopMode)matrix[i];
+            boolean isEntityStopped = (Boolean)matrix[i+1];
+            boolean expected = (Boolean)matrix[i+2];
+            boolean canStop = MachineLifecycleEffectorTasks.canStop(mode, isEntityStopped);
+            assertEquals(canStop, expected);
```
Comment (sjcorbett) [inline software/base/src/test/java/brooklyn/entity/basic/SoftwareProcessEntityTest.java]: Might be useful to add a message to the assert indicating the mode if we expect this to fail. It's overkill for this test but you could also use TestNG's [data providers](http://testng.org/doc/documentation-main.html#parameters-dataproviders).

File: software/base/src/main/java/brooklyn/entity/basic/SoftwareProcess.java [hunk 6]
```diff
@@ -261,6 +261,20 @@ private ChildStartableMode(boolean isDisabled, boolean isBackground, boolean isL
...
+
+        @Beta /** @since 0.7.0 semantics of parameters to restart being explored */
+        public static final ConfigKey<StopMode> STOP_PROCESS_MODE = ConfigKeys.newConfigKey(StopMode.class, "stopProcessMode",
+                "When to stop the process with regard to the entity state", StopMode.IF_NOT_STOPPED);
+
+        @Beta /** @since 0.7.0 semantics of parameters to restart being explored */
+        public static final ConfigKey<StopMode> STOP_MACHINE_MODE = ConfigKeys.newConfigKey(StopMode.class, "stopMachineMode",
+                "When to stop the machine with regard to the entity state. " +
+                "ALWAYS will try to stop the machine even if the entity is already stopped," +
+                "IF_NOT_STOPPED stops the machine only if entity not already stopped, " +
```
Comment (sjcorbett) [inline software/base/src/main/java/brooklyn/entity/basic/SoftwareProcess.java]: if the entity is

**Comment by sjcorbett** [comment]
Looks great. A few minor comments. Confirmed that unit and integration tests pass.

File: software/base/src/main/java/brooklyn/entity/software/MachineLifecycleEffectorTasks.java (hunk 4 above)
Comment (neykov) [inline software/base/src/main/java/brooklyn/entity/software/MachineLifecycleEffectorTasks.java]: The method is used in a test from a different package, that's why the public visibility is needed.

I changed the code a bit - the method is now protected with the test code providing a public wrapper, no need for the annotation. wdyt?

File: software/base/src/test/java/brooklyn/entity/basic/SoftwareProcessEntityTest.java (hunk 5 above)
Comment (neykov) [inline software/base/src/test/java/brooklyn/entity/basic/SoftwareProcessEntityTest.java]: Thanks for the data providers pointer, changed the code to use them. Logs the failing mode parameters automatically.

**Comment by neykov** [comment]
Addressed comments.

**Comment by sjcorbett** [comment]
I think your assertions are the wrong way round. Did you mean to `assertEquals(canStop, !expected)`?

File: software/base/src/test/java/brooklyn/entity/software/MachineLifecycleEffectorTasksTest.java [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+            { StopMode.IF_NOT_STOPPED, false, true },
+            { StopMode.NEVER, true, false },
+            { StopMode.NEVER, false, false },
+        };
+    }
+
+    @Test(dataProvider = "canStopStates")
+    public void testBasicSoftwareProcessCanStop(StopMode mode, boolean isEntityStopped, boolean expected) {
+        boolean canStop = canStop(mode, isEntityStopped);
+        assertEquals(canStop, !expected);
```
Comment (aledsage) [inline software/base/src/test/java/brooklyn/entity/software/MachineLifecycleEffectorTasksTest.java]: I agree with @sjcorbett and checked by running test: this should be `assertEquals(canStop, expected)`.

**Comment by aledsage** [comment]
@neykov Looks good; needs the test fixed in `MachineLifecycleEffectorTasksTest.java` and then can be merged.

**Comment by neykov** [comment]
Fixed test.

**Comment by neykov** [comment]
Not clear what went wrong with the last build - all projects are SUCCESS. Rebasing to force another build.

**Pull request merged**